File: backend/src/services/paymentService.js
// src/services/paymentService.js
// Passo 5 do fluxo: Financeiro valida e paga com fundos do cliente, dentro
// dos 7 dias. O Financeiro executa — não decide (a decisão já foi tomada
// pelo Company Admin no passo 2).

const { v4: uuid } = require('uuid');
const prisma = require('../config/database');
const { NotFoundError, ConflictError, ForbiddenError, ValidationError } = require('../utils/errors');
const notificationService = require('./notificationService');
const eventBus = require('./eventBus');
const platformFeeService = require('./platformFeeService');
const storageService = require('./storageService');
const auditService = require('./auditService');
const faturacaoService = require('./faturacaoService');
const agtSandboxSubmissionService = require('./agtSandboxSubmissionService');
const agtPayloadService = require('./agtPayloadService');
const logger = require('../config/logger');

async function listPendingInvoices(buyerCompanyId) {
  return prisma.invoice.findMany({
    where: {
      status: 'PENDENTE',
      OR: [
        { purchaseOrder: { buyerCompanyId } },
        { contract: { clientCompanyId: buyerCompanyId } },
      ],
    },
    include: { purchaseOrder: true, contract: true },
    orderBy: { dueAt: 'asc' },
  });
}

async function listPaymentHistory(buyerCompanyId) {
  return prisma.payment.findMany({
    where: {
      invoice: {
        OR: [
          { purchaseOrder: { buyerCompanyId } },
          { contract: { clientCompanyId: buyerCompanyId } },
        ],
      },
    },
    include: { invoice: { include: { purchaseOrder: true, contract: true } } },
    orderBy: { processedAt: 'desc' },
  });
}

async function processPayment(invoiceId, processedById, buyerCompanyId, proofFile = null, actor = null) {
  // Comprovativo da transferência OBRIGATÓRIO — é a prova (visível ao
  // fornecedor) de que o dinheiro saiu; sem ela o "pago" seria só uma palavra.
  if (!proofFile) {
    throw new ValidationError('Anexe o comprovativo da transferência (PDF ou imagem) para confirmar o pagamento.');
  }

  const invoice = await prisma.invoice.findUnique({
    where: { id: invoiceId },
    include: { purchaseOrder: true, contract: true },
  });
  if (!invoice) throw new NotFoundError('Fatura');
  if (invoice.status !== 'PENDENTE') {
    throw new ConflictError(`Fatura no estado "${invoice.status}" não pode ser paga.`);
  }

  const ownerCompanyId = invoice.purchaseOrder?.buyerCompanyId ?? invoice.contract?.clientCompanyId;
  if (ownerCompanyId !== buyerCompanyId) {
    throw new ForbiddenError('Só pode pagar faturas da sua própria empresa.');
  }

  // Guarda o comprovativo antes da transação (upload não é transacional).
  const proofUrl = await storageService.saveFile({
    buffer: proofFile.buffer,
    originalname: proofFile.originalname,
    mimetype: proofFile.mimetype,
    keyHint: `comprovativo-${invoice.reference}`,
    folder: 'proofs',
  });

  // O pagamento É o documento "RC" (Recibo) da AGT — série e cadeia de
  // integridade próprias do MESMO fornecedor que emitiu a fatura (ver
  // faturacaoService.serieReciboDoFornecedor).
  const supplierCompanyId = invoice.purchaseOrder?.supplierCompanyId ?? invoice.contract?.supplierCompanyId;
  const supplierCompany = supplierCompanyId
    ? await prisma.company.findUnique({
      where: { id: supplierCompanyId },
      select: { serieFiscal: true, dataAdesaoFacturacaoElectronica: true },
    })
    : null;

  const [payment] = await prisma.$transaction(async (tx) => {
    const certificacao = await faturacaoService.atribuir(tx, {
      emitidaEm: new Date(),
      total: invoice.amount,
      codigo: faturacaoService.serieReciboDoFornecedor(supplierCompany),
      dataAdesao: supplierCompany?.dataAdesaoFacturacaoElectronica,
    });

    const createdPayment = await tx.payment.create({
      data: {
        ...certificacao,
        invoiceId,
        amount: invoice.amount,
        currency: invoice.currency,
        processedById,
        reference: `PAY-${uuid().slice(0, 8).toUpperCase()}`,
        status: 'PROCESSADO',
        proofUrl,
        proofName: proofFile.originalname || 'comprovativo',
      },
    });

    await tx.invoice.update({ where: { id: invoiceId }, data: { status: 'PAGA' } });

    if (invoice.purchaseOrderId) {
      await tx.purchaseOrder.update({
        where: { id: invoice.purchaseOrderId },
        data: { status: 'PAGA', paidAt: new Date() },
      });
    } else if (invoice.consolidatedPoIds?.length) {
      await tx.purchaseOrder.updateMany({
        where: { id: { in: invoice.consolidatedPoIds } },
        data: { paidAt: new Date() },
      });
    }

    // Taxa da plataforma (KIXIMA) — à parte da PO/Fatura, cobrada ao fornecedor.
    const supplierCompanyId = invoice.purchaseOrder?.supplierCompanyId ?? invoice.contract?.supplierCompanyId;
    if (supplierCompanyId) {
      await platformFeeService.createForInvoice(tx, { invoice, companyId: supplierCompanyId });
    }

    // Auditoria DENTRO da transação: um pagamento sem registo não existe.
    await auditService.record(tx, {
      actor: actor || { actorId: processedById },
      action: 'PAGAMENTO_EXECUTADO',
      entityType: 'Payment',
      entityId: createdPayment.id,
      entityRef: createdPayment.reference,
      detail: {
        fatura: invoice.reference,
        valor: String(invoice.amount),
        moeda: invoice.currency,
        comprovativo: proofFile.originalname || 'comprovativo',
      },
    });

    return [createdPayment];
  });

  if (invoice.purchaseOrderId) {
    const po = await prisma.purchaseOrder.findUnique({ where: { id: invoice.purchaseOrderId } });
    await notificationService.events.pagamentoProcessado(payment, po);
  }

  // Evento para a integração ERP (pagamento concluído) — não-bloqueante.
  await eventBus.publish('payment.completed', eventBus.payloads.paymentCompleted(payment, invoice), {
    eventId: `payment-completed:${payment.id}`,
    tenantId: invoice.purchaseOrder?.buyerCompanyId ?? invoice.contract?.clientCompanyId ?? null,
  });

  // agtInvoiceResubmission — pedido explícito: ao confirmar o pagamento,
  // reenviar o FT desta fatura à AGT (registarFactura), com visibilidade do
  // payload e da resposta real (mesmo princípio de
  // agtSeriesService.solicitarSerie) — DIFERENTE da submissão RC acima
  // (agtSandboxSubmissionService, silenciosa, nunca lança). O FT já foi
  // submetido uma vez na emissão da fatura (poService.js/contractService.js)
  // — isto é um reenvio deliberado, não a primeira submissão; a AGT pode
  // recusá-lo por documentNo repetido, e é exatamente isso que se quer ver
  // aqui, não esconder. NUNCA bloqueia nem desfaz o pagamento já comitado
  // acima: o dinheiro já saiu, uma recusa de paperwork não desfaz isso.
  let agtInvoiceResubmission = null;
  if (supplierCompanyId) {
    try {
      const { payload, resposta, estado } = await agtPayloadService.submeterFatura(invoiceId, supplierCompanyId);
      agtInvoiceResubmission = { sucesso: true, payload, resposta, estado };
    } catch (erro) {
      agtInvoiceResubmission = {
        sucesso: false,
        erro: { message: erro.message, code: erro.code || null, details: erro.details || null },
      };
    }

    // Grava o resultado da ÚLTIMA tentativa na própria fatura — sem isto só
    // existia na resposta HTTP deste pedido, perdida depois. Numa falha de
    // escrita aqui (nunca esperada: a fatura já existe), só regista em log —
    // não pode desfazer nem falhar um pagamento já confirmado.
    try {
      await prisma.invoice.update({
        where: { id: invoiceId },
        data: agtInvoiceResubmission.sucesso
          ? {
            agtRequestId: agtInvoiceResubmission.resposta?.requestID ?? null,
            agtResultCode: agtInvoiceResubmission.resposta?.resultCode != null ? String(agtInvoiceResubmission.resposta.resultCode) : null,
            agtErro: null,
            // Estado real do processamento (obterEstado), já consultado
            // automaticamente por submeterFatura() assim que há requestID —
            // sem isto só existia na resposta HTTP deste pedido.
            agtEstado: agtInvoiceResubmission.estado ?? null,
          }
          : {
            // A AGT pode atribuir requestID mesmo numa recusa (ex.:
            // `errorList` preenchida mas ainda assim com requestID) —
            // guarda-se sempre que vier, é o que permite consultar depois o
            // motivo real via obterEstado (agtPayloadService.consultarEstadoFatura).
            agtRequestId: agtInvoiceResubmission.erro?.details?.respostaBruta?.requestID ?? null,
            agtResultCode: agtInvoiceResubmission.erro?.details?.resultCode != null ? String(agtInvoiceResubmission.erro.details.resultCode) : null,
            agtErro: agtInvoiceResubmission.erro,
            agtEstado: agtInvoiceResubmission.erro?.details?.estado ?? null,
          },
      });
    } catch (erroGravar) {
      logger.error('Falha ao gravar o resultado do reenvio AGT na fatura', { invoiceId, message: erroGravar.message });
    }

    await agtSandboxSubmissionService.submeter('RC', payment.id, supplierCompanyId);
  }

  return { ...payment, agtInvoiceResubmission };
}

// O fornecedor confirma que o valor entrou na conta — fecha o ciclo de
// confiança do "pagamento garantido". Uma vez por pagamento.
async function confirmReceived(paymentId, user, actor = null) {
  const payment = await prisma.payment.findUnique({
    where: { id: paymentId },
    include: { invoice: { include: { purchaseOrder: true, contract: true } } },
  });
  if (!payment) throw new NotFoundError('Pagamento');

  const supplierCompanyId =
    payment.invoice.purchaseOrder?.supplierCompanyId ?? payment.invoice.contract?.supplierCompanyId;
  if (supplierCompanyId !== user.companyId) {
    throw new ForbiddenError('Só o fornecedor desta fatura pode confirmar a receção.');
  }
  if (payment.receivedAt) {
    throw new ConflictError('A receção deste pagamento já foi confirmada.');
  }

  // Confirmação + auditoria na mesma transação (operação de dinheiro).
  const [updated] = await prisma.$transaction(async (tx) => {
    const upd = await tx.payment.update({
      where: { id: paymentId },
      data: { receivedAt: new Date(), receivedById: user.id },
    });
    await auditService.record(tx, {
      actor: actor || { actorId: user.id, companyId: user.companyId },
      action: 'RECECAO_VALOR_CONFIRMADA',
      entityType: 'Payment',
      entityId: payment.id,
      entityRef: payment.reference,
      detail: { fatura: payment.invoice.reference, valor: String(payment.amount), moeda: payment.currency },
    });
    return [upd];
  });
  return updated;
}

module.exports = { listPendingInvoices, listPaymentHistory, processPayment, confirmReceived };

File: backend-java/src/main/java/ao/kixima/payment/PaymentService.java
package ao.kixima.payment;

import ao.kixima.agt.AgtPayloadService;
import ao.kixima.agt.AgtSandboxSubmissionService;
import ao.kixima.audit.Actor;
import ao.kixima.audit.AuditService;
import ao.kixima.common.error.AppException;
import ao.kixima.common.error.ConflictException;
import ao.kixima.common.error.ForbiddenException;
import ao.kixima.common.error.NotFoundException;
import ao.kixima.common.error.ValidationException;
import ao.kixima.company.Company;
import ao.kixima.company.CompanyRepository;
import ao.kixima.faturacao.FaturacaoService;
import ao.kixima.invoice.Invoice;
import ao.kixima.invoice.InvoiceRepository;
import ao.kixima.invoice.InvoiceStatus;
import ao.kixima.invoice.dto.InvoiceDto;
import ao.kixima.messaging.EventBus;
import ao.kixima.messaging.EventPayloads;
import ao.kixima.notification.NotificationService;
import ao.kixima.payment.dto.PaymentDto;
import ao.kixima.po.PoStatus;
import ao.kixima.po.PurchaseOrder;
import ao.kixima.security.CurrentUser;
import ao.kixima.storage.StorageService;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.PlatformTransactionManager;
import org.springframework.transaction.support.TransactionTemplate;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * Espelha backend/src/services/paymentService.js — passo 5 do fluxo: o
 * Financeiro valida e paga com fundos do cliente. Executa, não decide (a
 * decisão foi do Company Admin no passo 2). O comprovativo é OBRIGATÓRIO:
 * sem ele o "pago" seria só uma palavra.
 *
 * Transações, como no Node: o pagamento nasce numa transação própria
 * (recibo certificado + fatura/PO PAGA + taxa da plataforma + auditoria);
 * o que vem DEPOIS (notificação, evento ERP, reenvio do FT à AGT, RC à
 * sandbox) nunca bloqueia nem desfaz um pagamento já comitado — o dinheiro
 * já saiu, uma recusa de paperwork não desfaz isso.
 *
 * NÃO PORTADO: o ramo {@code contract}/{@code consolidatedPoIds} (call-offs).
 */
@Service
public class PaymentService {

    private static final Logger log = LoggerFactory.getLogger(PaymentService.class);
    private static final long COMPROVATIVO_TAMANHO_MAXIMO = 10L * 1024 * 1024;

    private final PaymentRepository paymentRepository;
    private final InvoiceRepository invoiceRepository;
    private final CompanyRepository companyRepository;
    private final StorageService storageService;
    private final FaturacaoService faturacaoService;
    private final PlatformFeeService platformFeeService;
    private final AuditService auditService;
    private final NotificationService notificationService;
    private final EventBus eventBus;
    private final AgtPayloadService agtPayloadService;
    private final AgtSandboxSubmissionService agtSandboxSubmissionService;
    private final ObjectMapper objectMapper;
    private final TransactionTemplate tx;

    public PaymentService(PaymentRepository paymentRepository, InvoiceRepository invoiceRepository,
                           CompanyRepository companyRepository, StorageService storageService, FaturacaoService faturacaoService,
                           PlatformFeeService platformFeeService, AuditService auditService,
                           NotificationService notificationService, EventBus eventBus, AgtPayloadService agtPayloadService,
                           AgtSandboxSubmissionService agtSandboxSubmissionService, ObjectMapper objectMapper,
                           PlatformTransactionManager transactionManager) {
        this.paymentRepository = paymentRepository;
        this.invoiceRepository = invoiceRepository;
        this.companyRepository = companyRepository;
        this.storageService = storageService;
        this.faturacaoService = faturacaoService;
        this.platformFeeService = platformFeeService;
        this.auditService = auditService;
        this.notificationService = notificationService;
        this.eventBus = eventBus;
        this.agtPayloadService = agtPayloadService;
        this.agtSandboxSubmissionService = agtSandboxSubmissionService;
        this.objectMapper = objectMapper;
        this.tx = new TransactionTemplate(transactionManager);
    }

    public List<InvoiceDto> listPendingInvoices(String buyerCompanyId) {
        return tx.execute(s -> invoiceRepository.findPendentesDoComprador(buyerCompanyId, InvoiceStatus.PENDENTE).stream()
                .map(i -> InvoiceDto.de(i, true)).toList());
    }

    public List<PaymentDto> listPaymentHistory(String buyerCompanyId) {
        return tx.execute(s -> paymentRepository.findHistoricoDoComprador(buyerCompanyId).stream()
                .map(p -> PaymentDto.de(p, InvoiceDto.de(p.getInvoice(), true), null)).toList());
    }

    private record FaturaAPagar(String reference, String purchaseOrderId, String buyerCompanyId, String supplierCompanyId,
                                String serieFiscal, Instant dataAdesao) {
    }

    public PaymentDto processPayment(String invoiceId, CurrentUser user, MultipartFile proof, Actor actor) {
        if (proof == null || proof.isEmpty()) {
            throw new ValidationException("Anexe o comprovativo da transferência (PDF ou imagem) para confirmar o pagamento.");
        }
        if (proof.getSize() > COMPROVATIVO_TAMANHO_MAXIMO) throw new ValidationException("O comprovativo é demasiado grande (máximo 10 MB).");
        String tipo = proof.getContentType();
        boolean valido = tipo != null && (tipo.matches("^image/(png|jpe?g|webp|gif)$") || tipo.equals("application/pdf"));
        if (!valido) throw new ValidationException("Documento inválido — use PDF ou imagem (PNG/JPG).");

        FaturaAPagar fatura = tx.execute(s -> {
            Invoice invoice = invoiceRepository.findById(invoiceId).orElseThrow(() -> new NotFoundException("Fatura"));
            if (invoice.getStatus() != InvoiceStatus.PENDENTE) {
                throw new ConflictException("Fatura no estado \"" + invoice.getStatus() + "\" não pode ser paga.");
            }
            PurchaseOrder po = invoice.getPurchaseOrder();
            String ownerCompanyId = po == null ? null : po.getBuyerCompanyId();
            if (ownerCompanyId == null || !ownerCompanyId.equals(user.companyId())) {
                throw new ForbiddenException("Só pode pagar faturas da sua própria empresa.");
            }
            String supplierId = po.getSupplierCompanyId();
            Company supplier = supplierId == null ? null : companyRepository.findById(supplierId).orElse(null);
            return new FaturaAPagar(invoice.getReference(), invoice.getPurchaseOrderId(), ownerCompanyId, supplierId,
                    supplier == null ? null : supplier.getSerieFiscal(),
                    supplier == null ? null : supplier.getDataAdesaoFacturacaoElectronica());
        });

        // Guarda o comprovativo antes da transação (o upload não é transacional).
        String proofUrl;
        try {
            proofUrl = storageService.saveFile(proof.getBytes(), proof.getOriginalFilename(), tipo, "comprovativo-" + fatura.reference(), "proofs");
        } catch (IOException e) {
            throw new IllegalStateException("Falha a ler o comprovativo enviado.", e);
        }
        String proofName = proof.getOriginalFilename() == null || proof.getOriginalFilename().isBlank() ? "comprovativo" : proof.getOriginalFilename();

        Payment payment = tx.execute(s -> {
            Invoice invoice = invoiceRepository.findById(invoiceId).orElseThrow(() -> new NotFoundException("Fatura"));
            Instant agora = Instant.now();
            // O pagamento É o documento "RC" da AGT — série e cadeia próprias do MESMO fornecedor que emitiu a fatura.
            FaturacaoService.Certificacao certificacao = faturacaoService.atribuir(agora, invoice.getAmount(),
                    faturacaoService.serieReciboDoFornecedor(fatura.serieFiscal()), fatura.dataAdesao());

            Payment criado = new Payment(UUID.randomUUID().toString(), invoiceId, invoice.getAmount(), invoice.getCurrency(), user.id(),
                    "PAY-" + UUID.randomUUID().toString().substring(0, 8).toUpperCase(), proofUrl, proofName, agora, certificacao);
            paymentRepository.save(criado);

            invoice.setStatus(InvoiceStatus.PAGA);
            PurchaseOrder po = invoice.getPurchaseOrder();
            po.setStatus(PoStatus.PAGA);
            po.setPaidAt(agora);

            // Taxa da plataforma (KIXIMA) — à parte da PO/Fatura, cobrada ao fornecedor.
            if (fatura.supplierCompanyId() != null) platformFeeService.createForInvoice(invoice, fatura.supplierCompanyId());

            // Auditoria DENTRO da transação: um pagamento sem registo não existe.
            Map<String, Object> detail = new LinkedHashMap<>();
            detail.put("fatura", invoice.getReference());
            detail.put("valor", invoice.getAmount().toPlainString());
            detail.put("moeda", invoice.getCurrency());
            detail.put("comprovativo", proofName);
            auditService.record(new AuditService.Entry(actor != null ? actor : new Actor(user.id(), user.name(), null, user.companyId(), null),
                    "PAGAMENTO_EXECUTADO", "Payment", criado.getId(), criado.getReference(), detail));
            return criado;
        });

        Invoice invoiceComPo = tx.execute(s -> {
            Invoice i = invoiceRepository.findById(invoiceId).orElseThrow();
            i.getPurchaseOrder().getReference();
            return i;
        });
        notificationService.pagamentoProcessado(payment, invoiceComPo.getPurchaseOrder());

        eventBus.publish("payment.completed", EventPayloads.paymentCompleted(payment, invoiceComPo),
                "payment-completed:" + payment.getId(), fatura.buyerCompanyId());

        Object agtInvoiceResubmission = null;
        if (fatura.supplierCompanyId() != null) {
            agtInvoiceResubmission = reenviarFtAposPagamento(invoiceId, fatura.supplierCompanyId());
            agtSandboxSubmissionService.submeter("RC", payment.getId(), fatura.supplierCompanyId());
        }
        return PaymentDto.de(payment, null, agtInvoiceResubmission);
    }

    /**
     * Reenvio deliberado e VISÍVEL do FT à AGT ao confirmar o pagamento — a
     * AGT pode recusá-lo (documentNo repetido) e é exactamente isso que se
     * quer ver, não esconder. O resultado da ÚLTIMA tentativa fica na fatura.
     */
    private Map<String, Object> reenviarFtAposPagamento(String invoiceId, String supplierCompanyId) {
        Map<String, Object> resultado = new LinkedHashMap<>();
        String requestId = null;
        String resultCode = null;
        String erroJson = null;
        String estadoJson = null;
        try {
            AgtPayloadService.ResultadoSubmissao r = agtPayloadService.submeterFatura(invoiceId, supplierCompanyId);
            resultado.put("sucesso", true);
            resultado.put("payload", r.payload());
            resultado.put("resposta", r.resposta());
            resultado.put("estado", r.estado());
            Object rid = r.resposta() == null ? null : r.resposta().get("requestID");
            Object rc = r.resposta() == null ? null : r.resposta().get("resultCode");
            requestId = rid == null ? null : String.valueOf(rid);
            resultCode = rc == null ? null : String.valueOf(rc);
            estadoJson = jsonSilencioso(r.estado());
        } catch (Exception erro) {
            Map<String, Object> e = new LinkedHashMap<>();
            e.put("message", erro.getMessage());
            e.put("code", erro instanceof AppException ae ? ae.getCode() : null);
            e.put("details", erro instanceof AppException ae ? ae.getDetails() : null);
            resultado.put("sucesso", false);
            resultado.put("erro", e);
            erroJson = jsonSilencioso(e);
            if (erro instanceof AppException ae && ae.getDetails() instanceof ao.kixima.common.error.AgtRecusadoException.Details d) {
                resultCode = d.resultCode();
                if (d.respostaBruta() instanceof Map<?, ?> bruta && bruta.get("requestID") != null) requestId = String.valueOf(bruta.get("requestID"));
            }
        }

        try {
            final String fRequestId = requestId, fResultCode = resultCode, fErro = erroJson, fEstado = estadoJson;
            tx.executeWithoutResult(s -> invoiceRepository.findById(invoiceId).ifPresent(i -> {
                i.setAgtRequestId(fRequestId);
                i.setAgtResultCode(fResultCode);
                i.setAgtErro(fErro);
                i.setAgtEstado(fEstado);
            }));
        } catch (Exception erroGravar) {
            log.error("Falha ao gravar o resultado do reenvio AGT na fatura {}: {}", invoiceId, erroGravar.getMessage());
        }
        return resultado;
    }

    /** O fornecedor confirma que o valor entrou na conta — fecha o ciclo do "pagamento garantido". Uma vez por pagamento. */
    public PaymentDto confirmReceived(String paymentId, CurrentUser user, Actor actor) {
        return tx.execute(s -> {
            Payment payment = paymentRepository.findByIdComFatura(paymentId).orElseThrow(() -> new NotFoundException("Pagamento"));
            PurchaseOrder po = payment.getInvoice().getPurchaseOrder();
            String supplierCompanyId = po == null ? null : po.getSupplierCompanyId();
            if (supplierCompanyId == null || !supplierCompanyId.equals(user.companyId())) {
                throw new ForbiddenException("Só o fornecedor desta fatura pode confirmar a receção.");
            }
            if (payment.getReceivedAt() != null) throw new ConflictException("A receção deste pagamento já foi confirmada.");

            payment.confirmarRececao(Instant.now(), user.id());
            Map<String, Object> detail = new LinkedHashMap<>();
            detail.put("fatura", payment.getInvoice().getReference());
            detail.put("valor", payment.getAmount().toPlainString());
            detail.put("moeda", payment.getCurrency());
            auditService.record(new AuditService.Entry(actor != null ? actor : new Actor(user.id(), user.name(), null, user.companyId(), null),
                    "RECECAO_VALOR_CONFIRMADA", "Payment", payment.getId(), payment.getReference(), detail));
            return PaymentDto.de(payment, null, null);
        });
    }

    private String jsonSilencioso(Object valor) {
        if (valor == null) return null;
        try {
            return objectMapper.writeValueAsString(valor);
        } catch (Exception e) {
            return null;
        }
    }
}
